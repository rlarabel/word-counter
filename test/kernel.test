echo "Running tests..."
echo

make
./wc kernel.txt > TESTOUT

if [ $? -eq 0 ] ; then
  echo "Pass: Program exited zero"
else
  echo "Fail: Program did not exit zero"
  rm TESTOUT
  exit 1
fi

diff test/kernel.out TESTOUT > DIFFOUT

if [ $? -eq 0 ] ; then
  echo "Pass: Output is correct"
else
  echo "Fail: Output is not correct"
  cat DIFFOUT
  rm TESTOUT
  rm DIFFOUT
  exit 1
fi

echo
echo "All tests passed."

rm TESTOUT
rm DIFFOUT
exit 0

