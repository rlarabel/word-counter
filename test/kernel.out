In: 22
an: 44
operating: 65
system: 77
with: 56
a: 220
layered: 2
architecture,: 2
the: 439
kernel: 114
is: 142
level,: 4
has: 15
complete: 3
control: 4
of: 246
hardware: 24
and: 159
always: 3
in: 143
memory.: 4
some: 27
systems: 31
it: 48
single: 5
block: 2
memory,: 9
while: 8
other: 24
have: 24
mechanisms,: 2
e.g.,: 2
loadable: 2
that: 74
can: 38
kernel.: 16
The: 47
interactions: 2
between: 22
software: 9
A: 16
full: 3
controls: 2
all: 18
resources: 5
(e.g.: 2
via: 3
device: 25
drivers,: 2
processes: 13
such: 35
common: 9
e.g.: 2
CPU: 6
file: 12
systems,: 17
network: 6
On: 3
most: 17
one: 22
first: 8
programs: 16
loaded: 5
on: 34
startup: 3
It: 7
handles: 2
rest: 3
as: 75
well: 8
input/output: 2
requests: 9
from: 39
them: 5
into: 10
instructions: 3
for: 69
central: 2
processing: 2
unit.: 2
critical: 5
code: 21
usually: 6
separate: 8
area: 3
which: 45
protected: 5
access: 25
by: 50
application: 21
or: 50
less: 2
parts: 5
system.: 9
performs: 5
its: 11
running: 17
processes,: 4
devices: 11
hard: 2
handling: 2
this: 24
space.: 6
processors,: 2
video: 5
use: 25
user: 22
This: 28
separation: 5
prevents: 2
data: 14
each: 16
causing: 3
preventing: 2
applications: 14
crashing: 2
entire: 6
Even: 3
where: 11
included: 2
address: 19
spaces,: 2
memory: 47
protection: 23
used: 12
to: 260
prevent: 3
kernel's: 4
interface: 10
low-level: 3
abstraction: 6
layer.: 2
When: 7
process: 16
service: 4
kernel,: 11
must: 10
invoke: 2
through: 10
There: 3
are: 74
different: 13
architecture: 6
designs.: 3
Monolithic: 6
kernels: 46
run: 12
entirely: 2
space: 9
executing: 2
supervisor: 9
mode,: 3
mainly: 2
Microkernels: 3
but: 20
not: 37
their: 15
services: 11
space,: 3
like: 11
example: 6
microkernel: 22
design.: 2
Linux: 9
although: 3
also: 17
remove: 2
modules: 9
at: 27
runtime.: 2
computer: 13
responsible: 4
programs.: 5
responsibility: 3
deciding: 2
any: 6
time: 10
many: 12
should: 5
be: 65
processor: 5
==: 28
Random-access: 2
store: 2
both: 9
program: 16
data.: 2
need: 9
present: 2
order: 4
Often: 3
multiple: 3
will: 6
want: 2
frequently: 3
more: 21
than: 14
available.: 4
use,: 3
what: 11
do: 5
when: 8
enough: 2
I/O: 9
include: 5
peripherals: 2
disk: 5
printers,: 2
devices,: 2
display: 2
devices.: 3
perform: 9
appropriate: 2
provides: 5
methods: 2
using: 12
(typically: 2
abstracted: 2
point: 8
does: 7
implementation: 7
details: 3
management: 14
aspects: 2
necessary: 3
resource: 2
defining: 3
execution: 4
space): 2
mechanism: 9
within: 6
Kernels: 2
provide: 17
inter-process: 5
communication: 7
(IPC).: 2
These: 8
implementations: 2
may: 20
itself: 4
rely: 2
Although: 5
IPC: 4
provided: 11
method: 3
make: 6
these: 16
context: 4
switching: 3
system's: 2
allow: 7
safely: 2
they: 18
require: 6
doing: 2
virtual: 9
addressing,: 2
and/or: 3
Virtual: 5
addressing: 4
allows: 8
given: 2
physical: 2
another: 5
address,: 2
address.: 2
spaces: 2
accesses: 4
particular: 6
same: 10
every: 8
if: 11
only: 9
kernel): 3
thus: 3
currently: 3
would: 11
main: 3
As: 10
result,: 2
physically: 2
needs: 6
writing: 2
requested: 2
program.: 3
then: 8
was: 26
scheme: 3
generally: 6
known: 5
demand: 2
two: 6
being: 6
(kernel: 3
permitted: 4
fundamental: 2
much: 3
current: 5
designs: 3
actual: 2
general-purpose: 2
almost: 2
(e.g.,: 4
take: 3
Device: 3
To: 3
useful: 2
connected: 2
computer,: 2
drivers.: 3
driver: 9
enables: 2
interact: 2
device.: 2
information: 3
how: 3
communicate: 3
piece: 3
hardware.: 3
important: 4
design: 18
function: 6
translate: 2
abstract: 2
calls: 10
calls.: 6
work: 8
driver.: 2
drivers: 10
things: 2
cards,: 2
abstractions: 5
include:: 4
directly: 4
Using: 6
high-level: 5
drivers): 2
hardware,: 9
something: 3
direct: 3
Only: 3
primitives: 3
Implementing: 2
language: 6
example,: 4
request: 8
driver,: 2
actually: 3
maintain: 7
list: 5
available: 6
embedded: 4
rewritten: 3
PCs: 2
designed: 7
personal: 2
called: 14
very: 11
handled: 3
differently: 2
design,: 3
port: 3
decisions: 2
made: 3
system,: 10
making: 2
operation: 4
easily: 2
significant: 2
performance: 10
System: 5
call: 10
normally: 4
permission: 2
Most: 4
operations: 8
user-level: 3
performed: 4
requires: 5
They: 6
causes: 3
mode.: 6
An: 6
mode: 3
actions: 2
library: 7
normal: 2
C: 2
Windows: 14
passing: 5
read,: 2
wait: 2
able: 4
implemented: 9
API,: 2
impossible: 2
directly,: 2
because: 10
few: 2
are:: 4
therefore: 3
common.: 2
special: 5
stored: 3
location: 3
processor.: 2
instead: 3
without: 4
support,: 3
quite: 2
architectures: 3
been: 5
however,: 2
large: 5
result: 2
Kernel: 2
===: 34
support: 9
malicious: 2
clearly: 2
distinction: 2
leads: 2
hierarchical: 6
structure: 4
mechanisms: 6
policies: 5
according: 2
several: 5
static: 2
(enforced: 2
pre-emptive: 2
whether: 3
supported: 2
domains: 2
typically: 4
Many: 5
objects: 4
underlying: 2
object: 4
operations,: 2
including: 3
allowed: 4
read: 2
so: 3
checks: 3
Such: 2
applications.: 2
efficient: 5
simple: 3
way: 3
capabilities: 5
delegate: 2
access,: 2
capability-based: 3
commercial: 3
lack: 4
alternative: 3
approach: 16
simulate: 2
approach,: 3
capability: 3
(either: 2
delegating: 2
cost: 2
complex: 3
accessed: 2
firmware: 2
possible: 9
higher: 2
there: 6
language-based: 8
levels: 2
security: 15
(see: 3
build: 2
policy: 7
top: 4
features: 3
compiler.: 2
level: 6
often: 6
fact,: 5
support.: 3
====: 4
today: 2
about: 2
write: 2
spaces.: 2
execute: 3
trusted: 2
No: 2
Switching: 2
slow: 2
unnecessary: 2
switches: 3
systems.: 6
completely: 3
operate: 2
programming: 2
new: 5
time.: 2
verified: 2
started: 2
compiled: 2
correct: 2
recompiling: 2
source: 2
type: 2
traditional: 6
cannot: 3
means: 2
Process: 2
binary: 2
functionality: 4
terms: 2
safety: 2
message: 6
number: 5
approaches: 2
modern: 4
providing: 4
shared: 2
Brinch: 2
Hansen: 2
(although: 2
similar: 6
were: 14
this,: 4
allowing: 3
controller: 2
cause: 2
crash.: 2
With: 2
get: 4
program,: 3
Because: 2
manage: 2
done: 5
various: 3
(such: 3
active: 4
out: 5
tasks: 3
ways: 2
differ: 2
substantial: 2
philosophy: 2
micro: 9
monolithic: 36
kernels.: 10
policies,: 3
server: 4
minimal: 3
just: 2
basic: 5
remaining: 2
part: 8
management,: 4
high: 2
scheduling,: 2
tends: 2
them.: 2
presented: 2
properly: 2
major: 7
existing: 2
problem: 3
architecture.: 2
mode": 4
architectural: 3
conventional: 2
module: 6
"privileged: 2
issue: 2
fact: 2
together: 2
While: 2
space),: 2
microkernels: 11
modularity: 3
fit: 2
rather: 3
hybrid: 4
More: 3
exokernels: 2
production: 2
OS: 6
rich: 2
powerful: 2
Some: 5
UNIX: 2
implement: 3
microkernels.: 2
disadvantages: 3
-: 11
bug: 3
might: 4
crash: 2
become: 4
difficult: 4
maintain.: 2
kernels,: 5
traditionally: 2
Unix-like: 2
core: 3
functions: 4
contains: 2
put: 2
space:: 3
scheduler,: 3
handling,: 3
stacks.: 2
those: 7
general: 3
Modern: 2
(one: 2
GNU: 2
system): 2
feature: 2
ability: 4
load: 5
extension: 2
required,: 2
advantages: 3
Since: 2
smaller: 3
fewer: 2
bugs: 3
subsystem: 2
copy: 2
extremely: 4
small: 10
dynamically: 3
ease: 2
utilities: 2
still: 4
popular: 2
led: 3
types: 5
set: 4
servers: 3
over: 3
since: 2
totally: 2
integration: 2
bring: 2
down: 3
therefore,: 2
on.: 2
Examples: 2
term: 2
moved: 2
"user: 2
specific: 3
consists: 2
Other: 3
user-space: 2
programs,: 2
referred: 3
servers.: 2
easier: 2
overhead: 3
now: 4
whereby: 2
microkernel,: 6
(not: 2
necessarily: 2
QNX: 4
tested: 2
instance,: 3
development: 8
reboot: 2
general,: 2
basis: 2
microkernel.: 3
needs,: 2
MINIX,: 2
OS.: 2
required: 3
creates: 2
error: 2
However,: 3
whole: 4
networking: 3
modified: 2
starting: 2
For: 2
machine: 2
comparison: 2
potential: 2
due: 3
longer: 2
versus: 3
single-purpose: 2
(and: 2
among: 2
computing: 3
reduce: 2
lines: 2
code.: 2
early: 5
1990s,: 2
microkernels,: 2
considered: 3
virtually: 2
Andrew: 3
developers: 6
increase: 2
model: 3
based: 4
poor: 2
1980s: 2
did: 2
reasons: 2
left: 2
increased: 5
frequency: 3
manipulate: 3
Hybrid: 5
(or: 2
Microsoft: 4
NT: 6
Apple: 3
own: 2
macOS: 2
uses: 3
Mach: 3
kernel-space: 2
compromise: 2
implies: 2
adding: 4
modular: 3
essentially: 2
built: 2
people: 2
confused: 2
Faster: 2
modules.: 2
flexibility: 2
nanokernel: 2
even: 2
Exokernels: 3
no: 3
particularly: 2
computer.: 2
executed: 2
"bare: 2
metal": 2
computers: 5
during: 3
developed,: 2
became: 3
introduced: 2
could: 2
Unix,: 3
had: 2
One: 2
developments: 2
users: 5
time-sharing: 2
terminal: 2
Amiga: 2
released: 2
Unix: 6
programmers: 2
file,: 2
printed: 2
lower: 2
files: 2
dramatically: 2
series: 2
end: 2
utility: 2
up: 2
byte: 2
Linux,: 2
GNU,: 2
IBM: 3
FreeBSD,: 2
DragonflyBSD,: 2
Mac: 4
created: 3
line: 4
9x: 2
release: 3
developed: 3
regulates: 2
L4: 3
Operating: 3
Tanenbaum,: 2
Systems: 2
edition);: 2
Computer: 2

7929 kernel.txt

